storage/watchable_store_bench_test.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"math/rand"
+	"os"
+	"testing"
+	"time"
+)
+
+func BenchmarkWatchableStoreUnsyncedCancel(b *testing.B) {
+
+	watcherSize := b.N
+	if watcherSize == 1 {
+		// to not keep this test hanging for watcherSize 1
+		watcherSize++
+	}
+
+	s := newWatchableStore(tmpPath)
+
+	// To not call syncWatchersLoop and to make sure all watchers are in
+	// unsynced, this manually calls cleanup function.
+	close(s.stopc)
+	defer func() {
+		s.wg.Wait()
+		s.KV.Close()
+		os.Remove(tmpPath)
+	}()
+
+	b.ReportAllocs()
+
+	testKey := []byte("foo")
+	testValue := []byte("foo")
+	s.Put(testKey, testValue)
+
+	cancels := []CancelFunc{}
+	// create 2*N watchers
+	for i := 0; i < 2*watcherSize; i++ {
+		_, cancel := s.Watcher(testKey, true, -1)
+		cancels = append(cancels, cancel)
+	}
+
+	// random-cancel N watchers to make it not biased towards
+	// data structures with an order, such as slice.
+	rand.Seed(time.Now().UnixNano())
+	src := rand.NewSource(time.Now().UnixNano())
+	rd := rand.New(src)
+	idxSlice := rd.Perm(watcherSize)
+	idxSlice = idxSlice[:len(idxSlice)/2 : len(idxSlice)/2]
+
+	selectedCancels := []CancelFunc{}
+	for _, i := range idxSlice {
+		selectedCancels = append(selectedCancels, cancels[i])
+	}
+
+	b.ResetTimer()
+
+	// cancel N watchers
+	for _, cancel := range selectedCancels {
```

Comment:
idxSlice := rd.Perm(2*watcherSize)
// cancel N watchers
for _, idx := range idxSlice[:watcherSize] {
    cancels[idx]()
}